ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdListener.java

```diff
@@ -0,0 +1,109 @@
+package com.applovin.mediation;
+
+import android.content.Context;
+import android.util.Log;
+
+import com.applovin.nativeAds.AppLovinNativeAd;
+import com.applovin.nativeAds.AppLovinNativeAdLoadListener;
+import com.applovin.nativeAds.AppLovinNativeAdPrecacheListener;
+import com.applovin.sdk.AppLovinSdk;
+import com.applovin.sdk.AppLovinSdkUtils;
+import com.google.android.gms.ads.AdRequest;
+import com.google.android.gms.ads.mediation.MediationNativeListener;
+
+import java.lang.ref.WeakReference;
+import java.util.List;
+
+class AppLovinNativeAdListener
+        implements AppLovinNativeAdLoadListener, AppLovinNativeAdPrecacheListener {
+
+    private final AppLovinNativeAdapter mAdapter;
+    private final MediationNativeListener mNativeListener;
+    private final AppLovinSdk mAppLovinSdk;
+    private final WeakReference<Context> mContextWeakReference;
+
+    AppLovinNativeAdListener(AppLovinNativeAdapter adapter,
+                             MediationNativeListener nativeListener,
+                             AppLovinSdk sdk,
+                             Context context) {
+        mAdapter = adapter;
+        mNativeListener = nativeListener;
+        mAppLovinSdk = sdk;
+        mContextWeakReference = new WeakReference<>(context);
+    }
+
+    @Override
+    public void onNativeAdsLoaded(List<AppLovinNativeAd> nativeAds) {
+        if (nativeAds.size() > 0 && isValidNativeAd(nativeAds.get(0))) {
+            mAppLovinSdk.getNativeAdService().precacheResources(nativeAds.get(0), this);
+        } else {
+            Log.w(AppLovinNativeAdapter.TAG,
+                    "Ad from AppLovin doesn't have all assets required for the app install ad format");
+            notifyAdFailure(AdRequest.ERROR_CODE_NO_FILL);
+        }
+    }
+
+    @Override
+    public void onNativeAdsFailedToLoad(final int errorCode) {
+        notifyAdFailure(AppLovinUtils.toAdMobErrorCode(errorCode));
+    }
+
+    @Override
+    public void onNativeAdImagesPrecached(AppLovinNativeAd ad) {
+        // Create a native ad.
+        Context context = mContextWeakReference.get();
+        if (context == null) {
+            Log.w(AppLovinNativeAdapter.TAG, "Failed to create mapper. Context is null.");
+            notifyAdFailure(AdRequest.ERROR_CODE_INTERNAL_ERROR);
+            return;
+        }
+        final AppLovinNativeAdMapper mapper = new AppLovinNativeAdMapper(ad, context);
+        AppLovinSdkUtils.runOnUiThread(new Runnable() {
+            @Override
+            public void run() {
+                mNativeListener.onAdLoaded(mAdapter, mapper);
+            }
+        });
+    }
+
+    @Override
+    public void onNativeAdVideoPreceached(AppLovinNativeAd ad) {
+        // Do nothing.
+    }
+
+    @Override
+    public void onNativeAdImagePrecachingFailed(AppLovinNativeAd ad, final int errorCode) {
+        notifyAdFailure(AppLovinUtils.toAdMobErrorCode(errorCode));
```

Comment:
Can we log an error here? Basically we log all the callbacks, it's good for debugging and consistency (if you check out our other adapters).